subsys/bluetooth/common/log.c

```diff
@@ -40,18 +47,32 @@ const char *bt_hex_real(const void *buf, size_t len)
 
 const char *bt_addr_str_real(const bt_addr_t *addr)
 {
+#ifdef CONFIG_LOG_IMMEDIATE
+	static u8_t cur = 0;
+	static char bufs[4][BT_ADDR_STR_LEN];
+	char *str = bufs[cur];
+	cur = (cur + 1) % 4;
+#else
 	static char str[BT_ADDR_STR_LEN];
+#endif
 
-	bt_addr_to_str(addr, str, sizeof(str));
+	bt_addr_to_str(addr, str, BT_ADDR_STR_LEN);
 
 	return str;
 }
 
 const char *bt_addr_le_str_real(const bt_addr_le_t *addr)
 {
+#ifdef CONFIG_LOG_IMMEDIATE
+	static u8_t cur = 0;
+	static char bufs[4][BT_ADDR_LE_STR_LEN];
+	char *str = bufs[cur];
+	cur = (cur + 1) % 4;
```

Comment:
I wonder if we tap into the logging buffer to do such thing, though we don't free the memory anywhere but this is generating duplicated code as well.

Reply:
I guess you mean the log_strdup here? In which case log_strdup will return the same pointer under LOG_IMMEDIATE. Which is what is causing invalid (and confusing) logging output.

Reply:
The idea was to keep the log_strdup working even with LOG_IMMEDIATE that way we don't have to maintain 2 different code paths, though it appears we would be introducing a third way to log: https://github.com/zephyrproject-rtos/zephyr/pull/19322

Reply:
We could take this approach and just move it to the log_strdup function when using LOG_IMMEDIATE, instead of just returning the string like it does now.

Reply:
@Vudentz With minimal logging maybe we could get rid of LOG_IMMEDIATE, it seems to be problematic in many ways.

Reply:
Indeed, Id hope we do that so perhaps it is good idea to drop a comment in the PR.

Reply:
Already did. Let's hope they go for it.

Reply:
I guess one of the problem is the support of backends, Im not sure if the intention was not disable the backends as well but lets see their response.

Reply:
@Vudentz see my comment: https://github.com/zephyrproject-rtos/zephyr/pull/19322#issuecomment-534965754 imo, `LOG_MINIMAL` is not fixing issues with `LOG_IMMADIATE` but rather cuts features and reduces memory footprint.

Reply:
Well regardless what LOG_MINIMAL does all we need to do is to make log_strdup always dupping the string with is own buffers, that is fairly minimal and would work for any code using log_strdup with or without LOG_MINIMAL or LOG_IMMEDIATE.

Reply:
do i understand correctly, you meant that string duplicating feature should be present in minimal/immediate mode?

Reply:
Yes, absolutely Im not sure why it was thought not to be the case since it is quite obvious one could pass a local variable to log_strdup and then suddenly it doesn't work depending on the config options.